src/LayerPlanBuffer.cpp

```diff
@@ -107,6 +110,17 @@ void LayerPlanBuffer::addConnectingTravelMove(LayerPlan* prev_layer, const Layer
             // we force the path to use retraction
             path.retract = true;
         }
+        if (path.retract && !path.unretract_before_last_travel_move &&
+            next_extruder_settings.get<bool>("outer_inset_first") &&
+            (next_extruder_settings.get<coord_t>("infill_line_distance") == 0 ||
+                    !next_extruder_settings.get<bool>("infill_before_walls")))
+        {
+            // If there is a retraction and the next layer's first path is an outer wall, then we need to make sure that
+            // we unretract before the last travel move of the previous layer. This way we make sure that the unretract
+            // does not happen while the nozzle is on top of the new layer's outer wall, which could cause artefacts and
+            // decrease dimensional accuracy.
```

Comment:
There's a lot more to detecting if the next layer's first path is an outer wall... For instance, if there is any support, it will be printed before the outer wall. In fact, I've written a 160-line function once [here](https://github.com/Ultimaker/CuraEngine/blob/flow_advance/src/FffGcodeWriter.cpp#L3044) to determine which structure it should start with on the next layer. Things that could occur before an outer wall:
* Raft (not applicable)
* Prime tower
* Brim
* Skirt
* Ooze shield
* Draft shield
* Support infill
* Support roof
* Support floor
* Different meshes with per-mesh settings saying they have no walls, outer inset not first, etc.
* Surface mode

I'll do some testing to see if it actually goes wrong, but we might have a big simplification here of a complex problem.